### ISSUE TITLE
Issues when both extending and implementing in typescript

### ISSUE THREAD

**Issue by GeeWee** [prettier/prettier#3188]
**Prettier 1.8.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAdKcAeAHCAnGAAgEMAjAZxj2LCLABtjzzCBZCUzAFQE9s4AJgDF8AW0JZ4UASwAUIvOKaFghDAHdCsgJRISUHoQC+2wgEtR2enFEIYLERBXpUMAGYQnAXkKEA5KTEeH4A3OhG6CAANCAQ2DBm0OTIoEwwyG7E9ORwMaTUYADWcDAAytg0ZlAA5shUAK65IAIQYBlZOTEAVuSYAEIFxWXEtgAyVXDt2U0Q9TDYcwBMU50gFXg5eMggMHxw5GB4ZvHRa0ewAOpmAjAAFsgAHAAMMTmiZnV4ja9V1dYAivUIPAVk0YGQrjd7khFjEqMQzPRfgBhCCiUTEbZQaCTGL1HJcMjJJCZaZGIxAA)
```sh
--parser typescript
```

**Input:**
```tsx

export abstract class MobxTypedForm extends (Form as { new (): any }) implements Foo {
	foo =  'bar';
}

```

**Output:**
```tsx
export abstract class MobxTypedForm extends Form as { new (): any }
  implements Foo {
  foo = "bar";
}

```

**Expected behavior:**

I expect that the parenthesis around `(Form as { new (): any })` are not stripped away - as this leads to a compiler error.

**Comment by ikatyang** [comment]
**Second Output:**
```tsx
SyntaxError: ',' expected. (1:50)
> 1 | export abstract class MobxTypedForm extends Form as { new (): any }
    |                                                  ^
  2 |   implements Foo {
  3 |   foo = "bar";
  4 | }
```

**Comment by azz** [comment]
This has been fixed in `1.8.2`.
